### PR TITLE
Use dynamic filters to reduce hive partitions metadata listing

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description
Improves performance of splits generation on large partitioned hive tables by using 
dynamic filters to reduce the partitions metadata fetched by io.trino.metastore.HiveMetastore#getPartitionsByNames

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
## Hive
* Improve performance of queries with selective joins. ({issue}`23687`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java [hunk 1]
```diff
@@ -324,6 +331,15 @@ private Iterator<HivePartitionMetadata> getPartitionMetadata(
 
         Iterator<List<HivePartition>> partitionNameBatches = partitionExponentially(hivePartitions, minPartitionBatchSize, maxPartitionBatchSize);
         Iterator<List<HivePartitionMetadata>> partitionBatches = transform(partitionNameBatches, partitionBatch -> {
+            // Use dynamic filters to reduce the partitions listed by getPartitionsByNames
+            TupleDomain<ColumnHandle> currentDynamicFilter = dynamicFilter.getCurrentPredicate();
+            if (!currentDynamicFilter.isAll()) {
+                TupleDomain<ColumnHandle> partitionsFilter = currentDynamicFilter.intersect(tableHandle.getCompactEffectivePredicate());
+                partitionBatch = partitionBatch.stream()
+                        .filter(hivePartition -> partitionMatches(tableHandle.getPartitionColumns(), partitionsFilter, hivePartition))
```
Comment (Dith3r) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:339]: Do you have benchmark results?

**Review by Dith3r — APPROVED** [review]

**Review by sopel39 — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java (hunk 1 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:339]: Screenshot 2024-10-07 at 12 03 24 PM
[Hive DF partition metadata pruning.pdf](https://github.com/user-attachments/files/17277045/Hive.DF.partition.metadata.pruning.pdf)

Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:339]: I don't understand the results. Do they mean caching helps the most?

Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:339]: It means that this optimization helps when the metadata is not already in the cache. Catching of metadata is of course much better. We are not worse with metdata pruning + caching, there's just some duration flakiness on a few queries.

**Pull request merged**